src/compiler/checker.ts

```diff
@@ -28932,7 +28932,19 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         // and that call signature is non-generic, return statements are contextually typed by the return type of the signature
         const signature = getContextualSignatureForFunctionLikeDeclaration(functionDecl as FunctionExpression);
         if (signature && !isResolvingReturnTypeOfSignature(signature)) {
-            return getReturnTypeOfSignature(signature);
+            const returnType = getReturnTypeOfSignature(signature);
+            const functionFlags = getFunctionFlags(functionDecl);
+            if (functionFlags & FunctionFlags.Generator) {
+                return filterType(returnType, t => {
+                    return !!(t.flags & (TypeFlags.AnyOrUnknown | TypeFlags.Void | TypeFlags.InstantiableNonPrimitive)) || checkGeneratorInstantiationAssignabilityToReturnType(t, functionFlags, /*errorNode*/ undefined);

```

Comment:
I'm not super fond of this check here (talking mainly about `TypeFlags.InstantiableNonPrimitive` check, the rest is OK).

I analyzed all of the core logic behind this change and I think it's sound. This particular check might be slightly off though. In addition to that, I think that this logic could be improved further for instantiable types **but** I don't know how to do it right now. I think that this PR is an improvement already though and that it could be merged without fixing it right now.

I only have a single idea to fix this: create copies of instantiable types here with the same~ filtering logic applied to their constraints (perhaps the filtering could stay deferred so we wouldn't have to compute it until requested). I'm not sure how feasible this is and if it's the right approach so I'm not even taking a stab at this right now.

The overall idea behind this part of the fix is that we want to reject contextual types that are guaranteed to not match the given function's criteria (non-promises/non-generators). We also need to keep the types that poison the type (like any/unknown/void).

We need to do this here, for example, because `getContextualTypeForReturnExpression` uses the result of this function (`getContextualReturnType`). Let's consider this example:
```ts
declare function load(): Promise<boolean>;
type LoadCallback = () => Promise<boolean> | string;
const cb1: LoadCallback = async () => load().then(m => m);
```
In a case like this, we shouldn't allow `string` to be used as the contextual type for this return expression (of the inner arrow function). If we don't reject it soon enough then we end up with `Promise<string | boolean>` as the result of `getContextualTypeForReturnExpression` and thus we end up with an assignability problem as that's not assignable to `Promise<boolean>`.

Initially, I was trying to fix this by filtering in the `getContextualTypeForReturnExpression` [here](https://github.com/microsoft/TypeScript/pull/47683/files) but I found out that filtering in `getContextualReturnType` gives better results for some of the other added test cases in this PR here.